Fix a few issues with user commands not restarting game correctly

chris-mcdonald-dev opened the pull request:
I loved the refactor you started and decided to do some work on it as well.
Here's what's in this fix:

* **GAMES.JS / TYPING-GAME.JS** : 
  * Included a check for hangul commands when starting the typing game 
  * Added a function that checks if a user is trying to restart the game
    * This is to deal with users leaving a game unfinished and another user entering the channel and trying to start a new one sometime later.
  * Reordered the `clearTimeout()` and `resetGameVariables()` function calls in the `endGame()` function in typing-game.js
    * `resetGameVariables()` reset the `gameTimeout` to null, so `clearTimeout()` wasn't able to clear it previously.  

* **INDEX.JS** : 
  * Moved `gameIsRunning()` higher in the switch to allow `gameListener()` to check for when users try to restart the game.

Edit: _I merged the dev branch into your fork before I started working on it, so that's where the lines that look like additional changes to study sessions come from._

chris-mcdonald-dev committed:
Fix commands in typing game when user restarts game

Fox-Islam commented:
@chris-mcdonald-dev I didn't get notified about this at all! 😮 These changes look good! I just did a rebase to clean up the commits and the extra study session changes that were showing up in the commit. Happy to merge this into my branch if it looks good to you ✨

Fox-Islam committed:
Use regex for command matching

Tightens up the command matching a little bit instead of relying
on `includes` and `startsWith` but also helps make starting the game
a little bit easier, there are several instances of users trying
to start the game with !typing or @bot t or some other previously
invalid option

With this change, either a command prefix or a mention to the bot
should count as trying to use a command, so all of these work:
@bot typing
@bot !typing
!typing
@bot t
@bot !t
!t
@bot ㅌ
@bot !ㅌ
!ㅌ
